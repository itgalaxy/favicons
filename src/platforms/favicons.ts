import { FaviconHtmlElement } from "../index";
import { FaviconOptions, NamedIconOptions } from "../config/defaults";
import { transparentIcon, transparentIcons } from "../config/icons";
import { OptionalMixin, Platform, uniformIconOptions } from "./base";

const ICONS_OPTIONS: (NamedIconOptions & OptionalMixin)[] = [
  { name: "favicon.ico", ...transparentIcons(16, 24, 32, 48, 64) },
  { name: "favicon-16x16.png", ...transparentIcon(16) },
  { name: "favicon-32x32.png", ...transparentIcon(32) },
  { name: "favicon-48x48.png", ...transparentIcon(48) },
  { name: "favicon.svg", ...transparentIcon(1024), optional: true }, // arbitrary size. if more than one svg source is given, the closest to this size will be picked.
];

export class FaviconsPlatform extends Platform {
  constructor(options: FaviconOptions) {
    super(
      options,
      uniformIconOptions(options, options.icons.favicons, ICONS_OPTIONS),
    );
  }

  override async createHtml(): Promise<FaviconHtmlElement[]> {
    return this.iconOptions.map(({ name, ...options }) => {
      if (name.endsWith(".ico")) {
        // prettier-ignore
        return `<link rel="icon" type="image/x-icon" href="${this.cacheBusting(this.relative(name))}">`;
      } else if (name.endsWith(".svg")) {
        // prettier-ignore
        return `<link rel="icon" type="image/svg+xml" href="${this.cacheBusting(this.relative(name))}">`;
      }

      const { width, height } = options.sizes[0];

      // prettier-ignore
      return `<link rel="icon" type="image/png" sizes="${width}x${height}" href="${this.cacheBusting(this.relative(name))}">`;
    });
  }
}
