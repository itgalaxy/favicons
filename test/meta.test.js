import favicons from "../src";
import { logo_png } from "./util";

test("should allow specifying metadata", async () => {
  expect.assertions(1);

  const result = await favicons(logo_png, {
    appName: "PWA",
    appShortName: "PWA",
    appDescription: "Progressive Web App",
    developerName: "John Doe",
    developerURL: "https://john.doe.com",
    dir: "rtl",
    lang: "ar",
    background: "#333",
    theme_color: "#abc",
    appleStatusBarStyle: "default",
    display: "fullscreen",
    orientation: "portrait",
    scope: "/",
    start_url: "/subdomain/",
    version: "3.2.1",
  });

  await expect(result).toMatchFaviconsSnapshot();
});
