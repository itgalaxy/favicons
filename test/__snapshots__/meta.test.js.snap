// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow specifying metadata 63`] = `
Array [
  Object {
    "contents": null,
    "name": "favicon-16x16.png",
  },
  Object {
    "contents": null,
    "name": "favicon-32x32.png",
  },
  Object {
    "contents": null,
    "name": "favicon-48x48.png",
  },
  Object {
    "contents": null,
    "name": "favicon.ico",
  },
  Object {
    "contents": null,
    "name": "android-chrome-36x36.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-48x48.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-72x72.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-96x96.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-144x144.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-192x192.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-256x256.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-384x384.png",
  },
  Object {
    "contents": null,
    "name": "android-chrome-512x512.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-57x57.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-60x60.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-72x72.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-76x76.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-114x114.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-120x120.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-144x144.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-152x152.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-167x167.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-180x180.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-1024x1024.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-icon-precomposed.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-640x1136.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-750x1334.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-828x1792.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1125x2436.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1242x2208.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1242x2688.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1536x2048.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1668x2224.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1668x2388.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2048x2732.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1136x640.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2160x1620.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1620x2160.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1334x750.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-1792x828.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2436x1125.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2208x1242.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2688x1242.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2048x1536.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2224x1668.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2388x1668.png",
  },
  Object {
    "contents": null,
    "name": "apple-touch-startup-image-2732x2048.png",
  },
  Object {
    "contents": null,
    "name": "coast-228x228.png",
  },
  Object {
    "contents": null,
    "name": "firefox_app_60x60.png",
  },
  Object {
    "contents": null,
    "name": "firefox_app_128x128.png",
  },
  Object {
    "contents": null,
    "name": "firefox_app_512x512.png",
  },
  Object {
    "contents": null,
    "name": "mstile-70x70.png",
  },
  Object {
    "contents": null,
    "name": "mstile-144x144.png",
  },
  Object {
    "contents": null,
    "name": "mstile-150x150.png",
  },
  Object {
    "contents": null,
    "name": "mstile-310x150.png",
  },
  Object {
    "contents": null,
    "name": "mstile-310x310.png",
  },
  Object {
    "contents": null,
    "name": "yandex-browser-50x50.png",
  },
]
`;
